### DISCUSSION TITLE
use Entity Framework sql server

### DISCUSSION THREAD

**Discussion by guidoDimasi** [elsa-workflows/elsa-core#5382] [Q&A]
how i can use sql server as persistance layer?
how i can setup the migration to run automatically on startup?

**Comment by glime-ai** [comment]
To use SQL Server as the persistence layer for Elsa Workflows, you need to configure your application to use the `Elsa.Persistence.EntityFramework.SqlServer` package. Here's how you can do it:

1. Install the `Elsa.Persistence.EntityFramework.SqlServer` NuGet package in your project.

2. Configure your `Startup.cs` or `Program.cs` (depending on your .NET version) to use SQL Server with Elsa. Here's an example for .NET 6.0 or above:

```csharp
var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddElsa(elsa => elsa
    .UseEntityFrameworkPersistence(ef => ef.UseSqlServer(builder.Configuration.GetConnectionString("YourConnectionStringName")))
    // Add other Elsa services you might need
);

var app = builder.Build();

// Configure the HTTP request pipeline.
// ...

app.Run();
```

Make sure to replace `"YourConnectionStringName"` with the name of your connection string defined in your `appsettings.json` or other configuration source.

To set up the migration to run automatically on startup, you can create a hosted service that will apply any pending migrations to the database when the application starts. Here's an example of how you can do this:

```csharp
public class MigrationsHostedService : IHostedService
{
    private readonly IServiceProvider _serviceProvider;

    public MigrationsHostedService(IServiceProvider serviceProvider)
    {
        _serviceProvider = serviceProvider;
    }

    public async Task StartAsync(CancellationToken cancellationToken)
    {
        using var scope = _serviceProvider.CreateScope();
        var dbContext = scope.ServiceProvider.GetRequiredService<YourDbContext>(); // Replace with your actual DbContext type
        await dbContext.Database.MigrateAsync(cancellationToken);
    }

    public Task StopAsync(CancellationToken cancellationToken) => Task.CompletedTask;
}

// Register the hosted service in your `Startup.cs` or `Program.cs`
builder.Services.AddHostedService<MigrationsHostedService>();
```

Remember to replace `YourDbContext` with the actual type of your `DbContext`. This hosted service will ensure that migrations are applied at the start of your application.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Reply by guidoDimasi** [reply]
whats the type of the dbcontext used by elsa?
